File: python/erspy.cxx
#include <chrono>

#include <boost/noncopyable.hpp>
#include <boost/algorithm/string.hpp>
#include <boost/python.hpp>
#include <boost/python/exception_translator.hpp>

#include <mutex>
#include <unordered_map>

#include <ers/ers.h>
#include <ers/AnyIssue.h>
#include <ers/RemoteContext.h>

namespace
{
    struct Py_RH : boost::noncopyable {
	Py_RH(PyObject * object) : m_object(object)
	{ ; }

	~Py_RH()
	{ Py_DECREF(m_object); }

	operator PyObject *() const { return m_object; }

    private:
	PyObject * m_object;
    };


    PyObject * base_ex_type;
    
    PyObject * 
    to_python( PyObject *& custom_ex_type, ers::Issue const & ex )
    {
        static const std::string cpp_separator( "::" );
	static const std::string py_separator( "." );
	static std::unordered_map<std::string, PyObject *> ex_types;
	static std::mutex types_mutex;
        
        std::string str = ex.get_class_name();
        boost::replace_all( str, cpp_separator, py_separator );
        
        {
            std::unique_lock<std::mutex> lock(types_mutex);
            std::unordered_map<std::string, PyObject *>::iterator it = ex_types.find(str);
	    if (it == ex_types.end()) {
		custom_ex_type = PyErr_NewException( (char*)str.c_str(), base_ex_type, 0 );
		ex_types[str] = custom_ex_type;
	    } else {
		custom_ex_type = it->second;
	    }
        }

	Py_RH args(PyTuple_New( 3 ));
	PyTuple_SetItem( args, 0, PyUnicode_FromString(ex.what()) );
	PyTuple_SetItem( args, 1, PyDict_New() );
    	if ( !ex.cause() )
        {
    	    Py_INCREF( Py_None );
    	    PyTuple_SetItem( args, 2, Py_None );
        }
	else
	{
	    PyObject * t = 0;
            PyTuple_SetItem( args, 2, to_python( t, *ex.cause() ) );
        }

    	PyObject * e = PyObject_CallObject( custom_ex_type, args );

        PyObject_SetAttrString( e, "severity", PyLong_FromLong((ers::severity)ex.severity()) );
        
        const std::vector<std::string> & q = ex.qualifiers();
        Py_RH qualifiers(PyList_New(q.size()));
	for ( size_t i = 0; i < q.size(); ++i )
	{
            PyList_SetItem( qualifiers, i, PyUnicode_FromString( q[i].c_str() ) );
	}
        PyObject_SetAttrString( e, "qualifiers", qualifiers );
        
        const ers::string_map & p = ex.parameters();
        Py_RH parameters(PyObject_GetAttrString( e, "__dict__" ));
	for ( ers::string_map::const_iterator it = p.begin(); it != p.end(); ++it )
        {
            PyDict_SetItem( 	parameters, 
            			PyUnicode_FromString( it -> first.c_str() ), 
                                PyUnicode_FromString( it -> second.c_str() ) );
        }

	Py_RH c(PyObject_GetAttrString( e, "_Issue__context" ));
        PyObject_SetAttrString( c, "package_name", PyUnicode_FromString(ex.context().package_name()) );
        PyObject_SetAttrString( c, "file_name", PyUnicode_FromString(ex.context().file_name()) );
        PyObject_SetAttrString( c, "function_name", PyUnicode_FromString(ex.context().function_name()) );
        PyObject_SetAttrString( c, "line_number", PyLong_FromLong(ex.context().line_number()) );
        
        return e;
    }
    
    void 
    translate_exception( ers::Issue const & ex )
    {
        PyObject * t = 0;
        PyObject * e = to_python( t, ex );
        PyErr_SetObject( t, e );
        Py_DECREF( e );
    }

    void 
    init( PyObject * o )
    {
	base_ex_type = o;
    }
    
    ers::Issue * 
    issue( PyObject * o )
    {
        if (!o)
            return 0;

	std::string id	= PyBytes_AsString(
	    Py_RH( PyObject_GetAttrString( PyObject_Type( o ), "__name__" ) ) );
	
	std::string msg = PyBytes_AsString( Py_RH( PyObject_GetAttrString( o, "message" ) ) );
        
	Py_RH c(PyObject_GetAttrString( o, "context" ));
	ers::RemoteContext context(
		    PyBytes_AsString( Py_RH( PyObject_GetAttrString( c, "package_name" ) ) ),
		    PyBytes_AsString( Py_RH( PyObject_GetAttrString( c, "file_name" ) ) ),
		    PyLong_AsLong( Py_RH( PyObject_GetAttrString( c, "line_number" ) ) ),
		    PyBytes_AsString( Py_RH( PyObject_GetAttrString( c, "function_name" ) ) ),
		    ers::RemoteProcessContext(
			    PyBytes_AsString( Py_RH( PyObject_GetAttrString( c, "host_name" ) ) ),
			    PyLong_AsLong( Py_RH( PyObject_GetAttrString( c, "process_id" ) ) ),
			    PyLong_AsLong( Py_RH( PyObject_GetAttrString( c, "thread_id" ) ) ),
			    PyBytes_AsString( Py_RH( PyObject_GetAttrString( c, "cwd" ) ) ),
			    PyLong_AsLong( Py_RH( PyObject_GetAttrString( c, "user_id" ) ) ),
			    PyBytes_AsString( Py_RH( PyObject_GetAttrString( c, "user_name" ) ) ),
                            PyBytes_AsString( Py_RH( PyObject_GetAttrString( c, "application_name" ) ) ) ) );

        std::vector<std::string> qualifiers;
        Py_RH q(PyObject_GetAttrString( o, "qualifiers" ));
        int size = PyList_Size( q );
        for (int i = 0; i < size; ++i ) {
            qualifiers.push_back(PyBytes_AsString(PyList_GetItem(q, i)));
        }

	std::map<std::string, std::string> parameters;
	Py_RH items(PyDict_Items(Py_RH(PyObject_GetAttrString(o, "parameters"))));
        size = PyList_Size( items );
        for (int i = 0; i < size; ++i ) {
            PyObject * item = PyList_GetItem(items, i);
            parameters.insert( std::make_pair(
        	PyBytes_AsString(PyTuple_GetItem(item, 0)),
        	PyBytes_AsString(PyTuple_GetItem(item, 1))
        	));
        }

	double t = PyFloat_AsDouble( Py_RH( PyObject_GetAttrString(o, "time")));
	std::chrono::system_clock::time_point
	    time(std::chrono::nanoseconds(static_cast<int64_t>(t*1000000000.)));

	Py_RH cause(PyObject_GetAttrString(o, "cause"));

	return new ers::AnyIssue(id, ers::Error, context, time, msg,
				    qualifiers, parameters,
				    cause != Py_None ? issue( cause ) : 0 );
    }
    
    void
    debug( int level, PyObject * o )
    {
    	std::unique_ptr<ers::Issue> a( issue( o ) );
        ers::debug( *a, level );
    }
    
    void
    log( PyObject * o )
    {
    	std::unique_ptr<ers::Issue> a( issue( o ) );
    	ers::log( *a );
    }
    
    void
    info( PyObject * o )
    {
    	std::unique_ptr<ers::Issue> a( issue( o ) );
    	ers::info( *a );
    }
    
    void
    warning( PyObject * o )
    {
    	std::unique_ptr<ers::Issue> a( issue( o ) );
    	ers::warning( *a );
    }
    
    void
    error( PyObject * o )
    {
    	std::unique_ptr<ers::Issue> a( issue( o ) );
    	ers::error( *a );
    }
    
    void
    fatal( PyObject * o )
    {
    	std::unique_ptr<ers::Issue> a( issue( o ) );
    	ers::fatal( *a );
    }

    BOOST_PYTHON_MODULE(liberspy)
    {
	boost::python::register_exception_translator<ers::Issue>(&translate_exception);
	boost::python::def("init", &init);
	boost::python::def("debug", &debug);
	boost::python::def("log", &log);
	boost::python::def("info", &info);
	boost::python::def("warning", &warning);
	boost::python::def("error", &error);
	boost::python::def("fatal", &fatal);
    }
}

File: python/erspy.py
'''Lots of imports. May not need them all, prune list later'''
import sys
import ers
import inspect
import platform
import os
import sys
import re
import time
import getpass
import _thread
import logging

from ers import AnyIssue

    
''' Python exception class that can generate an ers isse'''            
class Issue( Exception ):
    "class for ERS exceptions"    
    __verbosity = int( os.getenv( "DUNEDAQ_ERS_VERBOSITY_LEVEL", "0" ) )
    
    def __init__( self, message, kwargs, cause ):
        Exception.__init__( self )
        #self.time = time.time()
        self.severity = ers.severity.Error
        self.message = message
        self.cause = cause
        self.__context = Context(  )
        self.qualifiers = self.__class__.__module__
        #self.qualifiers = [ self.__context.package_name ]
        self.__parameters = dict([ (str(k), str(v)) for (k,v) in list(kwargs.items()) ])
        self.__dict__.update( kwargs )
        
    @property
    def context( self ):
        return self.__context
            
    @property
    def parameters( self ):
        return self.__parameters
                    
    def isInstanceOf( self, cname ):
            return self.__class__.__name__ == cname and True or False
        
    def __repr__( self ):
        pretty_function = lambda f,v: v > 0 \
                                 and f \
                                 or  re.sub( '\(.*\)', '(...)', f )
        s = '%s %s [%s at %s:%d] %s' % ( 
                        SeverityNames[self.severity],
                        time.strftime( '%Y-%b-%d %H:%M:%S', time.localtime( self.time ) ),
                        pretty_function( self.__context.function_name, self.__verbosity ),
                        self.__context.file_name,
                        self.__context.line_number,
                        self.message )
        if self.cause != None:
            s += '\n\twas caused by: %s' % repr(self.cause)
        return s

    def __str__( self ):
        return self.message

    def issue( self ):
        # Create Context using the ers-python bindings
        rpc = ers.RemoteProcessContext(self.__context.host_name,self.__context.process_id,self.__context.thread_id,self.__context.cwd,self.__context.user_id,self.__context.user_name,self.__context.application_name)
        rc = ers.RemoteContext(self.qualifiers,self.__context.file_name,self.__context.line_number,self.__context.function_name,rpc)
        # Placeholder until I work out what would make most sense here
        type="Error"
        # Create and return an ers issue
        return ers.AnyIssue(type, rc, self.message)


'''Extract context information for creating an ers message '''
class Context( object ):
    "extracts context for an ERS issue"    
    __file = re.sub( r'\.pyc$', '.py', __file__ )

    def __init__( self ):
        self.stack = [f[0] for f in inspect.stack() \
                        if  f[1] != self.__file \
                        and (      'self' not in f[0].f_locals\
                                or not isinstance(f[0].f_locals['self'], Issue))]
                                
        class_name    = lambda : 'self' in self.stack[0].f_locals \
                                 and self.stack[0].f_locals['self'].__class__.__name__ + '.'\
                                 or ''

        self.function_name = class_name() + self.stack[0].f_code.co_name;
        try:
            self.function_name += inspect.formatargvalues( *inspect.getargvalues(self.stack[0]) )
        except Exception as exx:
            self.function_name += '(...)'
                
        self.file_name = self.stack[0].f_code.co_filename
        self.line_number = self.stack[0].f_lineno
        self.host_name = platform.node()
        self.cwd = os.getcwd()
        self.process_id = os.getpid()
        # Need to truncate this to fit 32 bit signed int
        self.thread_id = int(_thread.get_ident()%2**31)
        self.user_id = os.getuid()
        self.user_name = getpass.getuser()
        self.application_name = os.getenv( "DUNEDAQ_APPLICATION_NAME", "Undefined" )


'''Wrap a message string as an ERS issue'''
def Message(message):
    # This class collects information needed for message context
    ct = Context()
    # Still need to do something about these
    package_name = "erspy"
    type="Error"
    # Create Context using the ers-python bindings
    rpc = ers.RemoteProcessContext(ct.host_name,ct.process_id,ct.thread_id,ct.cwd,ct.user_id,ct.user_name,ct.application_name)
    rc = ers.RemoteContext(package_name,ct.file_name,ct.line_number,ct.function_name,rpc)
    # Create and return an ers issue
    return ers.AnyIssue(type, rc, message)


'''Bunch of functions to wrap a text string and inject it as an ers message of different types'''
def debug( msg, lvl ):
    "sends msg to the debug stream"
    ers.debug( isinstance( msg, ers.AnyIssue ) and msg or Message( msg ), lvl )

def log( msg ):
    "sends msg to the log stream"
    ers.log( isinstance( msg, ers.AnyIssue ) and msg or Message( msg ) )

def info( msg ):
    "sends msg to the information stream"
    ers.info( isinstance( msg, ers.AnyIssue ) and msg or Message( msg ) )

'''These just inject an issue into the more serious streams.
   The only thing these add over just calling the bound ers function directly is that they
   warn you if you use the wrong type of argument.
   Mainly here because the old ers.py had them, and they provide a consistency of interface
   rather than forcing the user to use these calls for some things and direct ers calls for others'''
def warning( issue ):
    "sends issue to the warning stream"
    assert isinstance(issue,ers.AnyIssue), \
            'Only an instance of ers.Issue sub-class can be sent to the ers.warning stream'
    ers.warning( issue )

def error( issue ):
    "sends issue to the error stream"
    assert isinstance(issue,ers.AnyIssue), \
            'Only an instance of ers.Issue sub-class can be sent to the ers.error stream'
    ers.error( issue )

def fatal( issue ):
    "sends issue to the fatal stream"
    assert isinstance(issue,ers.AnyIssue), \
            'Only an instance of ers.Issue sub-class can be sent to the ers.fatal stream'
    ers.fatal( issue )



class LoggingHandler( logging.Handler ) :
    severity_mapper = { logging.getLevelName( logging.DEBUG )   : lambda m: debug( m, 0 ),
                        logging.getLevelName( logging.INFO )    : info,
                        logging.getLevelName( logging.WARNING ) : warning,
                        logging.getLevelName( logging.ERROR )   : error,
                        logging.getLevelName( logging.CRITICAL ): fatal }
    def __init__(self):
        logging.Handler.__init__(self)
        
    def emit(self, record):
        if isinstance( record.msg, ers.AnyIssue ):
            self.severity_mapper[record.levelname]( record.msg )
        else:
            lc = ers.LocalContext(record.module,record.filename,record.lineno,record.funcName,False)
            #The original passes kwargs here as well as the message. Consider that for later
            message = str(record.msg)
            msg = ers.AnyIssue(record.levelname, lc, message)
            self.severity_mapper[record.levelname]( msg )

def addLoggingHandler( loggername = '' ):
    "adds ERS logging handler to the given logger"
    logging.getLogger( loggername ).addHandler( LoggingHandler() )

def replaceAllLoggingHandlers( loggername = '' ):
    """removes all logging handlers from the given logger 
    and adds ERS logging handler to it"""
    l = logging.getLogger( loggername )
    while len( l.handlers ) != 0:
        l.removeHandler(l.handlers[0])
    addLoggingHandler( loggername )
